chained/water_impl.go

```diff
@@ -42,54 +44,86 @@ func newWaterImpl(dir, addr string, pc *config.ProxyConfig, reportDialCore repor
 	ctx := context.Background()
 	wasmAvailableAt := ptSetting(pc, "water_available_at")
 	transport := ptSetting(pc, "water_transport")
-	if wasm == nil && wasmAvailableAt != "" {
-		vc := newWaterVersionControl(dir)
-		cli := httpClient
-		if cli == nil {
-			cli = proxied.ChainedThenDirectThenFrontedClient(1*time.Minute, "")
-		}
-		downloader, err := newWaterWASMDownloader(strings.Split(wasmAvailableAt, ","), cli)
-		if err != nil {
-			return nil, log.Errorf("failed to create wasm downloader: %w", err)
-		}
-
-		r, err := vc.GetWASM(ctx, transport, downloader)
-		if err != nil {
-			return nil, log.Errorf("failed to get wasm: %w", err)
-		}
-		defer r.Close()
-
-		b, err := io.ReadAll(r)
+	wg := new(sync.WaitGroup)
+	d := &waterImpl{
+		raddr:          addr,
+		reportDialCore: reportDialCore,
+		wgDownload:     wg,
+	}
+	if wasm != nil {
+		dialer, err := createDialer(ctx, wasm, transport)
 		if err != nil {
-			return nil, log.Errorf("failed to read wasm: %w", err)
-		}
-
-		if len(b) == 0 {
-			return nil, log.Errorf("received empty wasm")
+			return nil, log.Errorf("failed to create dialer: %w", err)
 		}
+		d.dialer = dialer
+	}
 
-		wasm = b
+	if wasm == nil && wasmAvailableAt != "" {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
```

Comment:
I've chosen to use a wait group because I can use it at tests to know when it loaded correctly the WASM file (when downloading). But it also makes sense to use a channel